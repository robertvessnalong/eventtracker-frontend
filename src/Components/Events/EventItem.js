import { useState, useContext } from 'react';
import {
  Card,
  Button,
  Container,
  Col,
  Form,
  Row,
  Alert,
} from 'react-bootstrap';
import UserContext from '../../Context/UserContext';
import EventFinderApi from '../../API/api';
import { AiFillHeart, AiOutlineHeart } from 'react-icons/ai';

const EventItem = ({ favorites, event }) => {
  const { user, setUser, loggedIn } = useContext(UserContext);
  const formInitalState = {
    eventId: event.id,
    eventName: event.title,
    userName: user.firstName,
    comment: '',
    uuid: user['uuid'],
  };
  const [formData, setFormData] = useState(formInitalState);
  const [comment, setIsCommenting] = useState(false);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);

  const handleCommentSection = () => {
    setIsCommenting(!comment);
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  // Handle Favorties
  const handleFavorites = async (e, event) => {
    const eventData = {
      type: 'event',
      eventId: event.id,
      eventName: event.title,
      eventDate: EventFinderApi.eventTimeConverter(event.datetime_utc),
      eventAddress: event.venue.address,
      eventExtendedAddress: event.venue.extended_address,
      eventUrl: event.url,
      uuid: user['uuid'],
    };
    const res = e.target.classList.contains('favorited')
      ? await EventFinderApi.removefavorite(eventData)
      : await EventFinderApi.favoriteEvent(eventData);
    if (res.error) {
      console.error(res.error);
    } else {
      const user = JSON.parse(localStorage.getItem('user'));
      const userData = await EventFinderApi.getUserInfo(user.user, user.token);
      setUser(userData);
    }
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await EventFinderApi.addComment(formData);
    if (res.error) {
      setError(true);
      setErrorMsg(res.error);
    } else if (res.comment) {
      const user = JSON.parse(localStorage.getItem('user'));
      const userData = await EventFinderApi.getUserInfo(user.user, user.token);
      setFormData(formInitalState);
      setIsCommenting(!comment);
      setUser(userData);
    }
  };

  return (
    <Card className='mt-4 mb-4 p-0'>
      <Card.Header as='h6'>
        <Row>
          <Col className='d-flex align-items-center' xs={10}>
            <small>{event.title}</small>
          </Col>
          <Col xs={2}>
            {loggedIn && (
              <>
                {EventFinderApi.favoriteCheck(favorites, event) ? (
                  <div
                    className='favorite-btn favorited'
                    onClick={(e) => handleFavorites(e, event)}
                  >
                    <AiFillHeart />
                  </div>
                ) : (
                  <div
                    className='favorite-btn'
                    onClick={(e) => handleFavorites(e, event)}
                  >
                    <AiOutlineHeart />
                  </div>
                )}
              </>
            )}
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        <Row>
          <Col xs={12} md={6} className='text-start'>
            <div className='card-time'>
              <b>{EventFinderApi.eventTimeConverter(event.datetime_utc)}</b>
            </div>
            <address>
              {event.venue.address}
              <br />
              {event.venue.extended_address}
            </address>
          </Col>
          <Col
            xs={12}
            md={6}
            className='d-flex align-items-center justify-content-end'
          >
            <a href={event.url} rel='noreferrer' target='_blank'>
              <Button className='blue-btn' variant='primary'>
                Buy Tickets
              </Button>
            </a>
          </Col>
        </Row>
      </Card.Body>
      <Card.Footer className='text-muted text-start'>
        <button onClick={handleCommentSection}>Comment</button>
        <Col className={comment ? `Comment Active` : `Comment`} xs={12}>
          <Container>
            {error && <Alert variant='danger'>{errorMsg}</Alert>}
            <h6>Comment On Event</h6>
            <Form onSubmit={handleSubmit}>
              <Form.Control
                as='textarea'
                placeholder='Leave a comment here'
                name='comment'
                style={{ height: '150px' }}
                value={formData.comment}
                onChange={handleChange}
              />
              <button type='Submit'>Submit</button>
            </Form>
          </Container>
        </Col>
      </Card.Footer>
    </Card>
  );
};

export default EventItem;
