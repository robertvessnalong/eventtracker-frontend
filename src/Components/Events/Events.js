import { useState, useEffect, useContext } from 'react';
import {
  Container,
  Row,
  Col,
  Spinner,
  FloatingLabel,
  Form,
  Button,
  Alert,
} from 'react-bootstrap';
import UserContext from '../../Context/UserContext';
import EventsFinderApi from '../../API/api';
import EventItem from './EventItem';
import uuid from 'react-uuid';

const Events = () => {
  const formInitalState = {
    search: '',
  };
  const { user } = useContext(UserContext);
  const [page, setPage] = useState(1);
  const [formData, setFormData] = useState(formInitalState);
  const [events, setEvents] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);
  const [hasEvents, setHasEvents] = useState(false);

  useEffect(() => {
    const getEvents = async () => {
      const res = await EventsFinderApi.getEvents({ page });
      setEvents(res);
      setLoading(false);
    };

    getEvents();
  }, [page]);

  if (loading) {
    return (
      <Row>
        <Col style={{ textAlign: 'center' }}>
          <Spinner animation='border' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </Spinner>
        </Col>
      </Row>
    );
  }

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await EventsFinderApi.getEvents(formData);
    if (res.error) {
      setError(true);
      setErrorMsg(res.error);
    } else {
      if (events.length > 0) {
        setHasEvents(false);
        setFormData(formInitalState);
        setEvents(res);
      } else {
        setHasEvents(true);
      }
    }
  };

  const handlePagination = (e) => {
    if (e.target.classList.contains('btn')) {
      const attr = e.target.getAttribute('aria-label');
      if (attr === 'Previous' && page !== 1) {
        setPage((prevState) => prevState - 1);
      } else if (attr === 'Next' && page >= 1) {
        setPage((prevState) => prevState + 1);
      }
    }
  };

  return (
    <div className='Events'>
      <header>
        <h2>Events</h2>
        <span>Find Events Near You</span>
        <hr></hr>
      </header>
      <Container>
        <div className='Search'>
          <Form onSubmit={handleSubmit} className='mt-4'>
            {error && <Alert variant='danger'>{errorMsg}</Alert>}
            <Row className='g-2'>
              <Col md>
                <FloatingLabel
                  controlId='floatingInputGrid'
                  label='Search Events...'
                >
                  <Form.Control
                    type='text'
                    name='search'
                    value={formData.search}
                    onChange={handleChange}
                  />
                </FloatingLabel>
              </Col>
              <Col md={3}>
                <Button
                  type='submit'
                  className='h-100 w-100 blue-btn'
                  variant='primary'
                >
                  Find Events
                </Button>{' '}
              </Col>
            </Row>
          </Form>
        </div>
      </Container>
      <Container className='Event-Items'>
        {hasEvents && (
          <Alert className='mt-4' variant='warning'>
            Sorry, no events found! Try Again
          </Alert>
        )}
        <Row xs={1} xl={2}>
          {events.map((event) => (
            <Col key={uuid()}>
              <EventItem event={event} favorites={user.favorites} />
            </Col>
          ))}
        </Row>
        <Row>
          <Col>
            <div aria-label='pagination'>
              <ul
                className='pagination justify-content-center'
                key={uuid()}
                onClick={handlePagination}
              >
                <Button
                  disabled={page === 1 ? true : false}
                  className='page-link'
                  href='#'
                  aria-label='Previous'
                  key={uuid()}
                >
                  <span className='page-arrow' aria-hidden='true'>
                    &laquo;
                  </span>
                </Button>
                <Button
                  className='page-link'
                  href='#'
                  aria-label='Next'
                  key={uuid()}
                >
                  <span className='page-arrow' aria-hidden='true'>
                    &raquo;
                  </span>
                </Button>
              </ul>
            </div>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Events;
