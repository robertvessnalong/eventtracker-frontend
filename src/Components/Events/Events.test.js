import {
  render,
  screen,
  fireEvent,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Events from './Events';
import TestUser from '../../Testing/TestUser';
import UserContext from '../../Context/UserContext';

describe('Events Component', () => {
  afterEach(jest.clearAllMocks);
  // Smoke Test
  test('renders without crashing', async () => {
    render(
      <UserContext.Provider value={{ user: TestUser }}>
        <Events />
      </UserContext.Provider>
    );
    await waitFor(() => {
      const headerSpan = screen.getByText(/Find Events Near You/i);
      expect(headerSpan).toBeInTheDocument();
    });
  });
  // Smoke Test
  test('Alert Not Rendered', async () => {
    render(
      <UserContext.Provider value={{ user: TestUser }}>
        <Events />
      </UserContext.Provider>
    );
    await waitFor(() => {
      const alertMsg = screen.queryByText(/Sorry, no events found! Try Again/i);
      expect(alertMsg).not.toBeInTheDocument();
    });
  });

  test('fireSubmit', async () => {
    const { getByText, getByRole } = render(
      <UserContext.Provider value={{ user: TestUser }}>
        <Events />
      </UserContext.Provider>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    await waitFor(() => {
      fireEvent.change(getByRole('textbox', { name: /search/i }), {
        query: '',
      });
      fireEvent.submit(getByText('Find Events'));
    });
    const alertMsg = screen.getByText(
      /The Great Jack OLantern Blaze - Long Island/i
    );
    expect(alertMsg).toBeInTheDocument();
  });

  // Snapshot
  test('matches snapshot', async () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: TestUser }}>
        <Events />
      </UserContext.Provider>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    expect(asFragment()).toMatchSnapshot();
  });
});
