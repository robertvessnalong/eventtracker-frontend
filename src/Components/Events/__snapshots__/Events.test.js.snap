// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Events"
  >
    <header>
      <h2>
        Events
      </h2>
      <span>
        Find Events Near You
      </span>
      <hr />
    </header>
    <div
      class="container"
    >
      <div
        class="Search"
      >
        <form
          class="mt-4"
        >
          <div
            class="g-2 row"
          >
            <div
              class="col-md"
            >
              <div
                class="form-floating"
              >
                <input
                  class="form-control"
                  id="floatingInputGrid"
                  name="search"
                  type="text"
                  value=""
                />
                <label
                  for="floatingInputGrid"
                >
                  Search Events...
                </label>
              </div>
            </div>
            <div
              class="col-md-3"
            >
              <button
                class="h-100 w-100 blue-btn btn btn-primary"
                type="submit"
              >
                Find Events
              </button>
               
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="Event-Items container"
    >
      <div
        class="row row-cols-xl-2 row-cols-1"
      >
        <div
          class="col"
        >
          <div
            class="mt-4 mb-4 p-0 card"
          >
            <h6
              class="card-header"
            >
              <div
                class="row"
              >
                <div
                  class="d-flex align-items-center col-10"
                >
                  <small>
                    The Great Jack OLantern Blaze - Long Island
                  </small>
                </div>
                <div
                  class="col-2"
                />
              </div>
            </h6>
            <div
              class="card-body"
            >
              <div
                class="row"
              >
                <div
                  class="text-start col-md-6 col-12"
                >
                  <div
                    class="card-time"
                  >
                    <b>
                      Sun Oct 24 2021 - 7 AM
                    </b>
                  </div>
                  <address>
                    1303 Round Swamp Road
                    <br />
                    Old Bethpage, NY 11804
                  </address>
                </div>
                <div
                  class="d-flex align-items-center justify-content-end col-md-6 col-12"
                >
                  <a
                    href="https://seatgeek.com/the-great-jack-olantern-blaze-long-island-tickets/family/2021-10-24-3-30-am/5491086"
                    rel="noreferrer"
                    target="_blank"
                  >
                    <button
                      class="blue-btn btn btn-primary"
                      type="button"
                    >
                      Buy Tickets
                    </button>
                  </a>
                </div>
              </div>
            </div>
            <div
              class="text-muted text-start card-footer"
            >
              <button>
                Comment
              </button>
              <div
                class="Comment col-12"
              >
                <div
                  class="container"
                >
                  <h6>
                    Comment On Event
                  </h6>
                  <form
                    class=""
                  >
                    <textarea
                      class="form-control"
                      name="comment"
                      placeholder="Leave a comment here"
                      style="height: 150px;"
                    />
                    <button
                      type="Submit"
                    >
                      Submit
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mt-5 row"
      >
        <div
          class="col"
        >
          <div
            aria-label="pagination"
          >
            <ul
              class="pagination justify-content-center"
            >
              <a
                aria-disabled="true"
                aria-label="Previous"
                class="page-link btn btn-primary disabled"
                role="button"
              >
                <span
                  aria-hidden="true"
                  class="page-arrow"
                >
                  «
                </span>
              </a>
              <a
                aria-label="Next"
                class="page-link btn btn-primary"
                href="#"
                role="button"
                tabindex="0"
              >
                <span
                  aria-hidden="true"
                  class="page-arrow"
                >
                  »
                </span>
              </a>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
